facebook/react-native
# Execute Animated.js declarative animation on UIThread on Android.

This is the first from the series of PRs I'm going to be sending shorty that would let Animated.js animations to run off the JS thread (for Android only).


This PR introduce a new native module that will be used for offloading animations - NativeAnimatedModule. It has a simple API that allows for animated nodes management via methods like: create/drop animated node, connect/disconnect nodes, start animation of a value node, attach/detach animated from a native view.


Similarly to how we handle UIManager view hierarchy updates we create a queue of animated graph operations that are then executed on the UI thread. This isolates us from problems that may be caused by concurrent updates of animated graph while UI thread is "executing" the animation.


The most important class NativeAnimatedNodesManager.java implements a management interface for animated nodes graph as well as implements a graph traversal algorithm that is run for each animation frame. For each animation frame we visit animated nodes that might've been updated as well as their children that may use parent's values to update themselves. At the end of the traversal algorithm we expect to reach a special type of the node: PropsAnimatedNode that is then responsible for calculating property map which can be sent to native view hierarchy to update the view.


From the "client" perspective the API is as simple as adding: `useNativeDriver: true` to the set of animation properties, e.g:
```
Animated.timing(
  this.state.fadeAnim,
  {
    toValue: 1,
    duration: 2000,
    useNativeDriver: true,
  },
).start();
```


This approach doesn't let us support animating layout properties as it updates the values directly on the native view hierarchy. In order to be able to handle updates of layout props we'd need to run the updates on shadow view hierarchy first (on "native" thread). However a lot of this code could be reused if we decided to implement support for layout props (my initial prototype was capable of updating layout props off the JS thread).


This PR adds support only for a limited number of Animated.js library features (mainly to keep it shorter), here is what's supported as of this PR:
 - Works only for android
 - Only one type of animation: Animated.timing
 - One can use animated values only for non-layout, numeric properties (e.g. support for transform property is disabled)
 - One can use JS driven version of Animated.js library along with "native" animated. The only limitation is that we cannot drive same value both by native and JS animated driver.


What's coming up next (have most of the work done but will be sending this later to keep this change as simple as possible):
 - More animation types: decay, spring
 - Support for transform property
 - Support for "operation" nodes: Animated.add, Animated.multiply, etc.
 - Animated.event support
 - Unit tests for the graph traversal algorithm


**IMPORTANT**: This PR doesn't enable animated native module by default. In order to use it you'd need to add `NativeAnimatedModule` to the list of yours app 'packages'


**Test plan**
 1. Add `NativeAnimatedModule` to the list of native modules for you app
 2. Add `useNativeDriver: true` to your animation config (Animated.timing only for now). You can use my example app: https://gist.github.com/kmagiera/bbfbac6c00077d188bc5
 3. Enable SPY_MODE in MessageQueue.js to see that setNativeProps operation are no longer passed over the bridge while animation is running and `useNativeDriver` is set to `true`.
